contracts/near/src/lib.rs

```diff
@@ -0,0 +1,489 @@
+//#![deny(warnings)]
+
+use {
+    error::Error,
+    ext::ext_wormhole,
+    near_sdk::{
+        borsh::{
+            self,
+            BorshDeserialize,
+            BorshSerialize,
+        },
+        collections::{
+            UnorderedMap,
+            UnorderedSet,
+        },
+        env,
+        is_promise_success,
+        log,
+        near_bindgen,
+        AccountId,
+        Balance,
+        BorshStorageKey,
+        Gas,
+        PanicOnDefault,
+        Promise,
+        StorageUsage,
+    },
+    p2w_sdk::{
+        BatchPriceAttestation,
+        Identifier,
+        PriceStatus,
+    },
+    state::{
+        PriceFeed,
+        Source,
+        Vaa,
+    },
+    std::io::Cursor,
+};
+
+pub mod error;
+pub mod ext;
+pub mod governance;
+pub mod state;
+
+const _GAS_FOR_VERIFY_VAA: Gas = Gas(10_000_000_000_000);
+const INITIAL_STALENESS_THRESHOLD: u64 = 60;
+
+#[derive(BorshSerialize, BorshStorageKey)]
+enum StorageKeys {
+    Source,
+    Prices,
+}
+
+/// The `State` contains all persisted state for the contract. This includes runtime configuration.
+///
+/// There is no valid Default state for this contract, so we derive PanicOnDefault to force
+/// deployment using one of the #[init] functions in the impl below.
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize, PanicOnDefault)]
+pub struct Pyth {
+    /// The set of `Source`s from which Pyth attestations are allowed to be relayed from.
+    pub sources: UnorderedSet<Source>,
```

Comment:
Dumb question, but do you mind explaining how access control works here? I.e. what prevents unauthenticated/unguarded modification of this state?

Reply:
The contract is invoked by the NEAR runtime, which means the only time an `&mut T` is made available is just before the runtime invokes it after a contract call. However, none of these need to be `pub` so I went ahead and removed them all anyway.